From 5c14845806ece71277615c089331cfc5d0f1470d Mon Sep 17 00:00:00 2001
From: Dan Nicholson <nicholson@endlessm.com>
Date: Wed, 16 Nov 2016 08:19:12 -0800
Subject: [PATCH] Kill chroot processes with SIGTERM before SIGKILL

Killing all processes in the chroot with SIGKILL does not give them a
chance to clean up. Currently this is causing an issue with fakeroot
where faked is started and creates a semaphore for IPC. When faked is
killed with SIGKILL, it can't clean up the semaphore. Eventually, all
semaphores are exhausted and faked can't be started on a subsequent
build.

Call killchroot once with SIGTERM and wait 2 seconds before calling with
SIGKILL. Hopefully this will give processes enough time to cleanup
before they're forcefully killed.

https://phabricator.endlessm.com/T14181
---
 build | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/build b/build
index 5534202..18db5f2 100755
--- a/build
+++ b/build
@@ -1092,6 +1092,10 @@ if test -n "$KILL" ; then
 	lxc-stop -n "$LXCID" || true
 	lxc-destroy -n "$LXCID"
     elif test -z "$VM_IMAGE" ; then
+	# Kill once with SIGTERM to allow processes to clean up and then
+	# follow with SIGKILL to ensure all processes are gone
+	$BUILD_DIR/killchroot -s 15 $BUILD_ROOT
+	sleep 2
 	if ! $BUILD_DIR/killchroot -s 9 $BUILD_ROOT ; then
 	    echo "could not kill build in $BUILD_ROOT"
 	    cleanup_and_exit 1
-- 
2.1.4

